astropy/io/fits/connect.py

```diff
@@ -272,30 +286,25 @@ def read_table_fits(
         # preserve null values.
         masked = mask = False
         fill_value = None
-        coltype = col.dtype.subdtype[0].type if col.dtype.subdtype else col.dtype.type
         if col.null is not None:
-            mask = data[col.name] == col.null
+            mask = arr == col.null
             # Return a MaskedColumn even if no elements are masked so
             # we roundtrip better.
             masked = True
             fill_value = col.null
         elif mask_invalid and issubclass(coltype, np.inexact):
```

Comment:
In principle, `np.issubdtype(arr.dtype, np.inexact)` would do the same thing, suggesting that `coltype` perhaps is not needed.